File: src/main/java/net/guizhanss/ultimategenerators2/implementation/items/generators/modular/ender_crystal/EnderCrystalGenerator.java
package net.guizhanss.ultimategenerators2.implementation.items.generators.modular.ender_crystal;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.EnderCrystal;
import org.bukkit.entity.EntityType;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.inventory.ItemStack;

import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup;
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack;
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType;
import io.github.thebusybiscuit.slimefun4.implementation.SlimefunItems;
import io.github.thebusybiscuit.slimefun4.implementation.operations.FuelOperation;
import io.github.thebusybiscuit.slimefun4.libraries.dough.blocks.BlockPosition;
import io.github.thebusybiscuit.slimefun4.utils.ChestMenuUtils;

import me.mrCookieSlime.CSCoreLibPlugin.Configuration.Config;
import me.mrCookieSlime.Slimefun.api.BlockStorage;
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu;
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenuPreset;

import net.guizhanss.ultimategenerators2.core.recipes.MachineFuel;
import net.guizhanss.ultimategenerators2.implementation.UGItems;
import net.guizhanss.ultimategenerators2.implementation.items.abstracts.AGenerator;
import net.guizhanss.ultimategenerators2.utils.Keys;
import net.guizhanss.ultimategenerators2.utils.LocationUtils;

public class EnderCrystalGenerator extends AGenerator {
    private static final Map<BlockPosition, Integer> CRYSTAL_COUNT = new HashMap<>();

    private static final int INFO_SLOT = 18;

    @ParametersAreNonnullByDefault
    protected EnderCrystalGenerator(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
        super(itemGroup, item, recipeType, recipe);
    }

    @Override
    protected void setup(@Nonnull BlockMenuPreset preset) {
        super.setup(preset);
        preset.addItem(INFO_SLOT, NOT_WORKING_ITEM, ChestMenuUtils.getEmptyClickHandler());
    }

    @Override
    public ItemStack getProgressBar() {
        return new ItemStack(Material.END_CRYSTAL);
    }

    @Override
    protected void registerDefaultFuelTypes() {
        registerFuel(7200, UGItems.RUNE_COMPLEX_ENDER, SlimefunItems.STONE_CHUNK);
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void onPlace(BlockPlaceEvent e, Block b) {
        super.onPlace(e, b);
        var loc = new BlockPosition(b);
        CRYSTAL_COUNT.put(loc, 0);
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void onBreak(BlockBreakEvent e, BlockMenu menu) {
        super.onBreak(e, menu);
        Location loc = menu.getLocation();
        CRYSTAL_COUNT.remove(new BlockPosition(loc));
        loc.getWorld().getNearbyEntities(loc, 4D, 4D, 4D, entity -> entity.getType() == EntityType.ENDER_CRYSTAL)
            .stream()
            .map(EnderCrystal.class::cast)
            .forEachOrdered(crystal -> {
                try {
                    if (!crystal.isShowingBottom()
                        && crystal.hasMetadata(Keys.ENHANCED_ENDER_CRYSTAL)
                        && (crystal.getBeamTarget().getBlock().getLocation().distance(LocationUtils.getBlockCenter(loc.add(0, -1, 0))) == 0D)) {
                        crystal.setBeamTarget(null);

                    }
                } catch (Exception ex) {
                    crystal.setBeamTarget(null);
                }
            });
    }

    public int getEnergyProduction(BlockPosition pos) {
        return super.getEnergyProduction() * CRYSTAL_COUNT.computeIfAbsent(pos, k -> 0);
    }

    @Override
    @ParametersAreNonnullByDefault
    public int getGeneratedOutput(Location l, Config data) {
        BlockMenu inv = BlockStorage.getInventory(l);
        BlockPosition pos = new BlockPosition(l);
        FuelOperation operation = processor.getOperation(l);

        // check for ender crystals
//        BlockPosition basePos = new BlockPosition(l.add(0, -1, 0));
//        AtomicInteger count = new AtomicInteger(0);
//        l.getWorld().getNearbyEntities(l, 4D, 4D, 4D, entity -> entity.getType() == EntityType.ENDER_CRYSTAL)
//            .stream()
//            .map(EnderCrystal.class::cast)
//            .forEachOrdered(crystal -> {
//                if (crystal.isShowingBottom() && crystal.getBeamTarget().getBlock().getLocation().distance(LocationUtils.getBlockCenter(basePos)) == 0D) {
//                    count.incrementAndGet();
//                    crystal.setBeamTarget(basePos.toLocation());
//                }
//            });
//        CRYSTAL_COUNT.put(pos, count.get());

        // check for fuel
        if (operation != null) {
            if (!operation.isFinished()) {
                processor.updateProgressBar(inv, STATUS_SLOT, operation);

                if (isChargeable()) {
                    int charge = getCharge(l, data);

                    if (getCapacity() - charge >= getEnergyProduction(pos)) {
                        operation.addProgress(1);
                        return getEnergyProduction(pos);
                    }

                    return 0;
                } else {
                    operation.addProgress(1);
                    return getEnergyProduction(pos);
                }
            } else {
                inv.replaceExistingItem(STATUS_SLOT, NOT_WORKING_ITEM);

                processor.endOperation(l);
                return 0;
            }
        } else if (CRYSTAL_COUNT.get(pos) > 0) {
            Map<Integer, Integer> found = new HashMap<>();
            MachineFuel fuel = findRecipe(inv, found);

            if (fuel != null) {
                for (Map.Entry<Integer, Integer> entry : found.entrySet()) {
                    inv.consumeItem(entry.getKey(), entry.getValue());
                }

                processor.startOperation(l, new FuelOperation(fuel.toSf()));
            }

            return 0;
        } else {
            return 0;
        }
    }
}

File: src/main/kotlin/net/guizhanss/ultimategenerators2/implementation/items/generators/modular/ender_crystal/EnderCrystalGenerator.kt
@file:Suppress("DEPRECATION")

package net.guizhanss.ultimategenerators2.implementation.items.generators.modular.ender_crystal

import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType
import io.github.thebusybiscuit.slimefun4.implementation.SlimefunItems
import io.github.thebusybiscuit.slimefun4.implementation.operations.FuelOperation
import io.github.thebusybiscuit.slimefun4.libraries.dough.blocks.BlockPosition
import io.github.thebusybiscuit.slimefun4.utils.ChestMenuUtils
import me.mrCookieSlime.CSCoreLibPlugin.Configuration.Config
import me.mrCookieSlime.Slimefun.api.BlockStorage
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenuPreset
import net.guizhanss.ultimategenerators2.implementation.UGItems
import net.guizhanss.ultimategenerators2.implementation.items.abstracts.AGenerator
import net.guizhanss.ultimategenerators2.utils.BlockUtils
import net.guizhanss.ultimategenerators2.utils.consts.Keys
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.EnderCrystal
import org.bukkit.entity.Entity
import org.bukkit.entity.EntityType
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.inventory.ItemStack

class EnderCrystalGenerator(
    itemGroup: ItemGroup,
    item: SlimefunItemStack,
    recipeType: RecipeType,
    recipe: Array<ItemStack?>
) : AGenerator(itemGroup, item, recipeType, recipe) {
    companion object {
        private val CRYSTAL_COUNT: MutableMap<BlockPosition, Int> = HashMap()

        private const val INFO_SLOT = 18
    }

    override fun setup(preset: BlockMenuPreset) {
        super.setup(preset)
        preset.addItem(INFO_SLOT, NOT_WORKING_ITEM, ChestMenuUtils.getEmptyClickHandler())
    }

    override val progressBar = ItemStack(Material.END_CRYSTAL)

    override fun registerDefaultFuelTypes() {
        registerFuel(7200, UGItems.RUNE_COMPLEX_ENDER, SlimefunItems.STONE_CHUNK)
    }

    override fun onPlace(e: BlockPlaceEvent, b: Block) {
        super.onPlace(e, b)
        val loc = BlockPosition(b)
        CRYSTAL_COUNT[loc] = 0
    }

    override fun onBreak(e: BlockBreakEvent, menu: BlockMenu) {
        super.onBreak(e, menu)
        val loc: Location = menu.location
        CRYSTAL_COUNT.remove(BlockPosition(loc))
        loc.world!!
            .getNearbyEntities(loc, 4.0, 4.0, 4.0) { entity: Entity -> entity.type == EntityType.ENDER_CRYSTAL }
            .stream()
            .map { obj: Entity -> obj as EnderCrystal }
            .forEachOrdered { crystal: EnderCrystal ->
                try {
                    if (!crystal.isShowingBottom
                        && crystal.hasMetadata(Keys.ENHANCED_ENDER_CRYSTAL)
                        && crystal.beamTarget!!.block.location
                            .distance(BlockUtils.getBlockCenter(loc.add(0.0, -1.0, 0.0))) == 0.0
                    ) {
                        crystal.beamTarget = null
                    }
                } catch (ex: Exception) {
                    crystal.beamTarget = null
                }
            }
    }

    private fun getEnergyProduction(pos: BlockPosition): Int {
        return super.energyProduction * CRYSTAL_COUNT.getOrPut(pos) { 0 }
    }

    override fun getGeneratedOutput(l: Location, data: Config): Int {
        val inv: BlockMenu = BlockStorage.getInventory(l)
        val pos = BlockPosition(l)
        val operation: FuelOperation? = processor.getOperation(l)

        // TODO: Implement this
        // check for ender crystals
//        BlockPosition basePos = new BlockPosition(l.add(0, -1, 0));
//        AtomicInteger count = new AtomicInteger(0);
//        l.getWorld().getNearbyEntities(l, 4D, 4D, 4D, entity -> entity.getType() == EntityType.ENDER_CRYSTAL)
//            .stream()
//            .map(EnderCrystal.class::cast)
//            .forEachOrdered(crystal -> {
//                if (crystal.isShowingBottom() && crystal.getBeamTarget().getBlock().getLocation().distance(LocationUtils.getBlockCenter(basePos)) == 0D) {
//                    count.incrementAndGet();
//                    crystal.setBeamTarget(basePos.toLocation());
//                }
//            });
//        CRYSTAL_COUNT.put(pos, count.get());

        // check for fuel
        if (operation != null) {
            if (!operation.isFinished) {
                processor.updateProgressBar(inv, STATUS_SLOT, operation)

                if (isChargeable) {
                    val charge = getCharge(l, data)

                    if (capacity - charge >= getEnergyProduction(pos)) {
                        operation.addProgress(1)
                        return getEnergyProduction(pos)
                    }

                    return 0
                } else {
                    operation.addProgress(1)
                    return getEnergyProduction(pos)
                }
            } else {
                inv.replaceExistingItem(STATUS_SLOT, NOT_WORKING_ITEM)

                processor.endOperation(l)
                return 0
            }
        } else if (CRYSTAL_COUNT.getValue(pos) > 0) {
            val found: MutableMap<Int, Int> = HashMap()
            val fuel = findRecipe(inv, found)

            if (fuel != null) {
                for ((key, value) in found) {
                    inv.consumeItem(key, value)
                }

                processor.startOperation(l, FuelOperation(fuel.toSf()))
            }

            return 0
        } else {
            return 0
        }
    }
}
